src/Adapter/Product/Image/Repository/ProductImageMultiShopRepository.php

```diff
@@ -0,0 +1,242 @@
+use PrestaShop\PrestaShop\Core\Domain\Shop\Exception\InvalidShopConstraintException;
+use PrestaShop\PrestaShop\Core\Domain\Shop\ValueObject\ShopConstraint;
+use PrestaShop\PrestaShop\Core\Domain\Shop\ValueObject\ShopId;
+use PrestaShop\PrestaShop\Core\Exception\CoreException;
+use PrestaShop\PrestaShop\Core\Repository\AbstractMultiShopObjectModelRepository;
+use PrestaShopException;
+
+/**
+ * Provides access to product Image data source with shop context
+ */
+class ProductImageMultiShopRepository extends AbstractMultiShopObjectModelRepository
+{
+    /**
+     * @var Connection
+     */
+    private $connection;
+
+    /**
+     * @var string
+     */
+    private $dbPrefix;
+    /**
+     * @var ProductMultiShopRepository
+     */
+    private $productMultiShopRepository;
+
+    /**
+     * @param Connection $connection
+     * @param string $dbPrefix
+     */
+    public function __construct(
+        Connection $connection,
+        string $dbPrefix,
+        ProductMultiShopRepository $productMultiShopRepository
+    ) {
+        $this->connection = $connection;
+        $this->dbPrefix = $dbPrefix;
+        $this->productMultiShopRepository = $productMultiShopRepository;
+    }
+
+    /**
+     * @param ProductId $productId
+     *
+     * @return Image[]
+     *
+     * @throws CoreException
+     */
+    public function getImages(ProductId $productId, ShopConstraint $shopConstraint): array
+    {
+        $imagesIds = $this->getImagesIds($productId, $shopConstraint);
+
+        if ($shopConstraint->getShopGroupId()) {
+            throw new InvalidShopConstraintException('Image has no features related with shop group use single shop and all shops constraints');
+        } elseif ($shopConstraint->forAllShops()) {
+            $shopId = $this->productMultiShopRepository->getProductDefaultShopId($productId);
+        } else {
+            $shopId = $shopConstraint->getShopId();
+        }
+
+        if (empty($imagesIds)) {
+            return [];
+        }
+
+        $images = [];
+        foreach ($imagesIds as $imageId) {
+            $images[] = $this->get($imageId, $shopId);
+        }
+
+        return $images;
+    }
+
+    /**
+     * @param ProductId $productId
+     *
+     * @return ImageId[]
+     *
+     * @throws CoreException
+     */
+    public function getImagesIds(ProductId $productId, ShopConstraint $shopConstraint): array
+    {
+        $qb = $this->connection->createQueryBuilder()->select('id_image');
+
+        //todo: manage position
+        if ($shopConstraint->getShopGroupId()) {
+            throw new InvalidShopConstraintException('Image has no features related with shop group use single shop and all shops constraints');
+        } elseif ($shopConstraint->forAllShops()) {
+            $qb->from($this->dbPrefix . 'image', 'i')
+                ->addOrderBy('i.position', 'ASC')
+            ;
+        } else {
+            $qb->from($this->dbPrefix . 'image_shop', 'i')
```

Comment:
Ah, actually my description above is incorrect since some filtering is performed here. But I thought we discussed about the expected behaviour of `GetProductImages` which will be used for the Images component on product page. In the page we will always need all the images. 

Now I realize this doesn't help with the behat tests but I believe you can still use the `ProductImage::getShopIds` method to check the association correctly even if all the images are always returned.